File: code_cpp/ch01/gaussian_elimination.cpp
//==============================================================================
// Chapter 1: Numerical Stability in Gaussian Elimination
// C++ Implementation with Growth Factor Analysis
//==============================================================================
#include <iostream>
#include <vector>
#include <Eigen/Dense>
#include <Eigen/LU>
#include <Eigen/QR>
#include <random>
#include <chrono>
#include <iomanip>
#include <algorithm>

using namespace Eigen;
using namespace std;

//==============================================================================
// 1. Gaussian Elimination with Partial Pivoting (Educational Implementation)
//==============================================================================
struct LUResult {
    MatrixXd L;
    MatrixXd U;
    VectorXi P;
    double growth_factor;
};

LUResult gaussian_elimination_partial_pivot(const MatrixXd& A_in) {
    int n = A_in.rows();
    MatrixXd A = A_in;
    MatrixXd L = MatrixXd::Identity(n, n);
    VectorXi P = VectorXi::LinSpaced(n, 0, n-1);
    
    double max_A = A.cwiseAbs().maxCoeff();
    double max_U = max_A;
    
    for (int k = 0; k < n-1; ++k) {
        // Find pivot
        int i_max = k;
        double max_val = std::abs(A(k, k));
        for (int i = k+1; i < n; ++i) {
            if (std::abs(A(i, k)) > max_val) {
                max_val = std::abs(A(i, k));
                i_max = i;
            }
        }
        
        if (i_max != k) {
            // Swap rows in A
            A.row(k).swap(A.row(i_max));
            // Swap rows in L (only previous columns)
            for (int j = 0; j < k; ++j) {
                std::swap(L(k, j), L(i_max, j));
            }
            // Track permutation
            std::swap(P(k), P(i_max));
        }
        
        // Eliminate
        for (int i = k+1; i < n; ++i) {
            L(i, k) = A(i, k) / A(k, k);
            for (int j = k+1; j < n; ++j) {
                A(i, j) -= L(i, k) * A(k, j);
            }
            max_U = max(max_U, std::abs(A(i, k+1)));
        }
    }
    
    LUResult result;
    result.L = L;
    result.U = A.template triangularView<Upper>();
    result.P = P;
    result.growth_factor = max_U / A_in.cwiseAbs().maxCoeff();
    return result;
}

//==============================================================================
// 2. Solve using LU factors
//==============================================================================
VectorXd solve_lu(const MatrixXd& L, const MatrixXd& U, const VectorXi& P, const VectorXd& b) {
    int n = L.rows();
    VectorXd Pb = b(P);  // Apply permutation
    
    // Forward substitution: L y = Pb
    VectorXd y = VectorXd::Zero(n);
    for (int i = 0; i < n; ++i) {
        y(i) = Pb(i) - (L.row(i).head(i)).dot(y.head(i));
    }
    
    // Backward substitution: U x = y
    VectorXd x = VectorXd::Zero(n);
    for (int i = n-1; i >= 0; --i) {
        x(i) = (y(i) - U.row(i).tail(n-i-1).dot(x.tail(n-i-1))) / U(i, i);
    }
    return x;
}

//==============================================================================
// 3. Wilkinson Matrix (Worst Case for Partial Pivoting)
//==============================================================================
MatrixXd wilkinson_matrix(int n) {
    MatrixXd A = MatrixXd::Identity(n, n);
    for (int i = 0; i < n; ++i) {
        for (int j = i+1; j < n; ++j) {
            A(j, i) = -1.0;
        }
    }
    A.col(n-1).setOnes();
    return A;
}

//==============================================================================
// 4. Growth Factor Computation
//==============================================================================
double compute_growth_factor(const MatrixXd& A) {
    LUResult result = gaussian_elimination_partial_pivot(A);
    return result.growth_factor;
}

//==============================================================================
// 5. Benchmark: Eigen vs Manual Implementation
//==============================================================================
void benchmark_gaussian_elimination() {
    cout << "=== Gaussian Elimination Benchmark ===\n\n";
    
    vector<int> sizes = {100, 500, 1000, 2000};
    
    cout << setw(8) << "Size" << setw(15) << "Eigen LU (ms)" 
         << setw(15) << "Manual (ms)" << setw(15) << "Speedup"
         << setw(15) << "Growth ρ" << "\n";
    cout << string(68, '-') << "\n";
    
    for (int n : sizes) {
        MatrixXd A = MatrixXd::Random(n, n);
        VectorXd b = VectorXd::Random(n);
        
        // Eigen's LU (highly optimized)
        auto t1 = chrono::high_resolution_clock::now();
        FullPivLU<MatrixXd> lu_eigen(A);
        VectorXd x_eigen = lu_eigen.solve(b);
        auto t2 = chrono::high_resolution_clock::now();
        double eigen_ms = chrono::duration<double, milli>(t2 - t1).count();
        
        // Manual implementation
        t1 = chrono::high_resolution_clock::now();
        LUResult result = gaussian_elimination_partial_pivot(A);
        VectorXd x_manual = solve_lu(result.L, result.U, result.P, b);
        t2 = chrono::high_resolution_clock::now();
        double manual_ms = chrono::duration<double, milli>(t2 - t1).count();
        
        // Verify accuracy
        double error = (x_eigen - x_manual).norm() / x_eigen.norm();
        
        cout << setw(8) << n << setw(15) << fixed << setprecision(2) << eigen_ms
             << setw(15) << manual_ms << setw(15) << setprecision(1) << manual_ms/eigen_ms
             << setw(15) << setprecision(2) << result.growth_factor << "\n";
    }
}

//==============================================================================
// 6. Exact vs Floating Point (using rational via GMP - placeholder)
//==============================================================================
void benchmark_exact_vs_float() {
    cout << "\n=== Exact vs Floating Point ===\n";
    cout << "Note: Exact rational arithmetic requires GMP/MPFR libraries.\n";
    cout << "This is a placeholder showing the concept.\n\n";
    
    int n = 100;
    MatrixXd A = MatrixXd::Random(n, n);
    VectorXd b = VectorXd::Random(n);
    
    // Float64
    auto t1 = chrono::high_resolution_clock::now();
    VectorXd x_float = A.fullPivLu().solve(b);
    auto t2 = chrono::high_resolution_clock::now();
    double float_ms = chrono::duration<double, milli>(t2 - t1).count();
    
    // "Exact" would use Rational<BigInt> - extremely slow
    cout << "n = " << n << "\n";
    cout << "  Float64 time: " << fixed << setprecision(2) << float_ms << " ms\n";
    cout << "  Exact rational: ~hours (not implemented here)\n";
    cout << "  Typical speedup factor: 10^4 - 10^6 x\n";
}

//==============================================================================
// 7. Wilkinson Matrix Test
//==============================================================================
void test_wilkinson_matrix() {
    cout << "\n=== Wilkinson Matrix (Worst Case) ===\n";
    
    for (int n : {10, 20, 30, 50, 100}) {
        MatrixXd W = wilkinson_matrix(n);
        double rho = compute_growth_factor(W);
        cout << "n = " << setw(3) << n << " : growth factor ρ = " 
             << scientific << setprecision(2) << rho 
             << " (theoretical max: 2^(n-1) = " << pow(2.0, n-1) << ")\n";
    }
}

//==============================================================================
// 8. Condition Number and Forward Error Analysis
//==============================================================================
void error_analysis_demo(int n = 100) {
    cout << "\n=== Error Analysis Demo ===\n";
    
    // Well-conditioned matrix
    MatrixXd A_well = MatrixXd::Random(n, n) + n*MatrixXd::Identity(n, n);
    double kappa_well = A_well.condition();
    
    // Ill-conditioned matrix
    MatrixXd A_ill = MatrixXd::Random(n, n);
    A_ill = A_ill * A_ill.transpose();  // symmetric positive definite
    A_ill(0,0) = 1e-12;  // Make nearly singular
    double kappa_ill = A_ill.condition();
    
    VectorXd b = VectorXd::Random(n);
    VectorXd x_true_well = A_well.fullPivLu().solve(b);
    VectorXd x_true_ill = A_ill.fullPivLu().solve(b);
    
    // Solve with Float64 (which uses partial pivoting via LAPACK)
    VectorXd x_f_well = A_well.cast<float>().fullPivLu().solve(b.cast<float>()).cast<double>();
    VectorXd x_f_ill = A_ill.cast<float>().fullPivLu().solve(b.cast<float>()).cast<double>();
    
    double err_well = (x_f_well - x_true_well).norm() / x_true_well.norm();
    double err_ill = (x_f_ill - x_true_ill).norm() / x_true_ill.norm();
    
    cout << "Well-conditioned (κ ≈ " << kappa_well << "):\n";
    cout << "  Relative error: " << err_well << "\n";
    cout << "Ill-conditioned (κ ≈ " << kappa_ill << "):\n";
    cout << "  Relative error: " << err_ill << "\n";
    
    // Growth factors
    double rho_well = compute_growth_factor(A_well);
    double rho_ill = compute_growth_factor(A_ill);
    cout << "\nGrowth factors: well= " << rho_well << ", ill= " << rho_ill << "\n";
}

//==============================================================================
// Main
//==============================================================================
int main() {
    cout.precision(6);
    
    benchmark_gaussian_elimination();
    test_wilkinson_matrix();
    benchmark_exact_vs_float();
    error_analysis_demo(100);
    
    cout << "\n=== Key Takeaways ===\n";
    cout << "1. Partial pivoting growth factor is typically O(n) for random matrices\n";
    cout << "2. Wilkinson matrix achieves exponential growth 2^(n-1)\n";
    cout << "3. Eigen's LU is 10-100x faster than naive implementation\n";
    cout << "4. Exact rational arithmetic is impractical for n > 50\n";
    cout << "5. Growth factor ρ = ||U||_max / ||A||_max predicts stability\n";
    cout << "6. Exact rational arithmetic is impractical for n > 50\n";
    
    return 0;
}

File: code/ch01/gaussian_elimination.jl
# Chapter 1: Numerical Stability in Gaussian Elimination
# Complete Julia implementation with growth factor analysis

using LinearAlgebra, Random, Statistics, Printf
using SparseArrays, Plots

# ============================================================================
# 1. Basic Gaussian Elimination (no pivoting) - for educational purposes
# ============================================================================
function gaussian_elimination_no_pivot(A::Matrix{Float64})
    """Naive Gaussian elimination without pivoting - DO NOT USE IN PRACTICE"""
    n = size(A, 1)
    A = copy(A)
    L = Matrix{Float64}(I, n, n)
    
    for k = 1:n-1
        if abs(A[k,k]) < eps()
            error("Zero pivot encountered at step $k")
        end
        for i = k+1:n
            L[i,k] = A[i,k] / A[k,k]
            for j = k+1:n
                A[i,j] -= L[i,k] * A[k,j]
            end
        end
    end
    U = UpperTriangular(A)
    return L, U
end

# ============================================================================
# 2. Gaussian Elimination with Partial Pivoting (standard)
# ============================================================================
function gaussian_elimination_partial_pivot(A::Matrix{Float64})
    """Gaussian elimination with partial (row) pivoting"""
    n = size(A, 1)
    A = copy(A)
    L = Matrix{Float64}(I, n, n)
    P = Vector{Int}(1:n)
    
    for k = 1:n-1
        # Find pivot row
        i_max = argmax(abs.(A[k:n, k])) + k - 1
        if i_max != k
            # Swap rows in A
            A[k, :], A[i_max, :] = A[i_max, :], A[k, :]
            # Swap rows in L (only previous columns)
            L[k, 1:k-1], L[i_max, 1:k-1] = L[i_max, 1:k-1], L[k, 1:k-1]
            # Track permutation
            P[k], P[i_max] = P[i_max], P[k]
        end
        
        for i = k+1:n
            L[i,k] = A[i,k] / A[k,k]
            for j = k+1:n
                A[i,j] -= L[i,k] * A[k,j]
            end
        end
    end
    U = UpperTriangular(A)
    return L, U, P
end

function solve_lu(L, U, P, b)
    """Solve PA = LU system: A x = b"""
    n = length(b)
    # Apply permutation
    Pb = b[P]
    # Forward substitution: L y = Pb
    y = zeros(n)
    for i = 1:n
        y[i] = Pb[i] - dot(L[i, 1:i-1], y[1:i-1])
    end
    # Backward substitution: U x = y
    x = zeros(n)
    for i = n:-1:1
        x[i] = (y[i] - dot(U[i, i+1:n], x[i+1:n])) / U[i,i]
    end
    return x
end

# ============================================================================
# 3. Growth Factor Computation
# ============================================================================
function growth_factor(A::Matrix{Float64}; pivoting=:partial)
    """Compute the growth factor ρ = max|U_ij| / max|A_ij|"""
    n = size(A, 1)
    if pivoting == :none
        L, U = gaussian_elimination_no_pivot(A)
    elseif pivoting == :partial
        L, U, P = gaussian_elimination_partial_pivot(A)
    else
        error("Unknown pivoting strategy")
    end
    max_A = maximum(abs, A)
    max_U = maximum(abs, U.data)
    return max_U / max_A
end

# ============================================================================
# 4. Wilkinson Matrix (worst case for partial pivoting)
# ============================================================================
function wilkinson_matrix(n::Int)
    """Wilkinson matrix - worst case for partial pivoting"""
    A = Matrix{Float64}(I, n, n)
    for i = 1:n
        for j = i+1:n
            A[j,i] = -1.0
        end
    end
    A[:,end] .= 1.0
    return A
end

# ============================================================================
# 5. Benchmark: Exact vs Floating Point
# ============================================================================
function benchmark_exact_vs_float(n=1000)
    """Compare exact rational vs Float64 solve time and accuracy"""
    Random.seed!(42)
    A = rand(Bool, n, n)
    b = rand(Bool, n)
    
    # Float64 solve
    A_f = Float64.(A)
    b_f = Float64.(b)
    t1 = @elapsed x_f = A_f \ b_f
    
    # Exact rational solve (using Rational{BigInt})
    A_r = Rational{BigInt}.(A)
    b_r = Rational{BigInt}.(b)
    t2 = @elapsed x_r = A_r \ b_r
    
    # Relative error
    x_f_big = BigFloat.(x_f)
    x_r_float = Float64.(x_r)
    rel_error = norm(x_f_big - x_r_float) / norm(x_r_float)
    
    println("n = $n")
    println("  Float64 time: $(round(t1*1000, digits=2)) ms")
    println("  Exact time:   $(round(t2, digits=2)) s")
    println("  Speedup:      $(round(t2/t1))x")
    println("  Relative error: $rel_error")
    
    return t1, t2, rel_error
end

# ============================================================================
# 6. Growth Factor Experiments
# ============================================================================
function growth_factor_experiment()
    """Replicate the empirical growth factor table from the article"""
    Random.seed!(123)
    println("Growth factors for random matrices (partial pivoting):")
    println("n\tmedian\t99th percentile")
    
    for n in [100, 500, 1000, 5000]
        rhos = Float64[]
        for trial = 1:100
            A = randn(n, n)
            push!(rhos, growth_factor(A, pivoting=:partial))
        end
        println("$n\t$(round(median(rhos), digits=1))\t$(round(quantile(rhos, 0.99), digits=1))")
    end
    
    # Wilkinson matrix
    println("\nWilkinson matrix (worst case):")
    for n in [20, 50, 100]
        A = wilkinson_matrix(n)
        rho = growth_factor(A, pivoting=:partial)
        println("n=$n: ρ = $rho (theoretical max: $(2^(n-1)))")
    end
end

# ============================================================================
# 7. Condition Number and Forward Error Analysis
# ============================================================================
function error_analysis_demo(n=100)
    """Demonstrate relationship between condition number, growth factor, and error"""
    Random.seed!(42)
    
    # Well-conditioned matrix
    A_well = randn(n, n) + n*I
    κ_well = cond(A_well)
    
    # Ill-conditioned matrix
    A_ill = randn(n, n)
    A_ill = A_ill * A_ill'  # symmetric positive definite
    A_ill[1,1] = 1e-12  # Make nearly singular
    κ_ill = cond(A_ill)
    
    b = randn(n)
    x_true_well = A_well \ b
    x_true_ill = A_ill \ b
    
    # Solve with Float64 (which uses partial pivoting via LAPACK)
    x_f_well = Float64.(A_well) \ Float64.(b)
    x_f_ill = Float64.(A_ill) \ Float64.(b)
    
    err_well = norm(x_f_well - x_true_well) / norm(x_true_well)
    err_ill = norm(x_f_ill - x_true_ill) / norm(x_true_ill)
    
    println("Well-conditioned (κ ≈ $(round(κ_well, digits=1))):")
    println("  Relative error: $err_well")
    println("Ill-conditioned (κ ≈ $(round(κ_ill, digits=1))):")
    println("  Relative error: $err_ill")
    
    # Growth factors
    rho_well = growth_factor(A_well, pivoting=:partial)
    rho_ill = growth_factor(A_ill, pivoting=:partial)
    println("\nGrowth factors: well= $rho_well, ill= $rho_ill")
end

# ============================================================================
# Main
# ============================================================================
if abspath(PROGRAM_FILE) == @__FILE__
    println("=== Chapter 1: Gaussian Elimination Stability ===\n")
    
    # Run growth factor experiment
    growth_factor_experiment()
    
    println("\n=== Error Analysis ===")
    error_analysis_demo(100)
    
    println("\n=== Benchmark (smaller for demo) ===")
    benchmark_exact_vs_float(200)
end